In [79]:
import seaborn as sns
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np

In [220]:
sharks = pd.read_csv("data/attacks.csv", encoding = 'ISO-8859-1')

sharks

Case Number         Date    Year        Type    Country  \
0      2018.06.25  25-Jun-2018  2018.0     Boating        USA   
1      2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA   
2      2018.06.09  09-Jun-2018  2018.0     Invalid        USA   
3      2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA   
4      2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO   
...           ...          ...     ...         ...        ...   
25718         NaN          NaN     NaN         NaN        NaN   
25719         NaN          NaN     NaN         NaN        NaN   
25720         NaN          NaN     NaN         NaN        NaN   
25721         NaN          NaN     NaN         NaN        NaN   
25722          xx          NaN     NaN         NaN        NaN   

                  Area                        Location     Activity  \
0           California     Oceanside, San Diego County     Paddling   
1              Georgia  St. Simon Island, Glynn County     Standing   
2               Hawaii                    Habush, Oahu      Surfing   
3      New South Wales              Arrawarra Headland      Surfing   
4               Colima                        La Ticla  Free diving   
...                ...                             ...          ...   
25718              NaN                             NaN          NaN   
25719              NaN                             NaN          NaN   
25720              NaN                             NaN          NaN   
25721              NaN                             NaN          NaN   
25722              NaN                             NaN          NaN   

                  Name Sex   ...         Species   \
0          Julie Wolfe    F  ...      White shark   
1      Adyson McNeely     F  ...              NaN   
2          John Denges    M  ...              NaN   
3                 male    M  ...        2 m shark   
4       Gustavo Ramos     M  ...  Tiger shark, 3m   
...                ...  ...  ...              ...   
25718              NaN  NaN  ...              NaN   
25719              NaN  NaN  ...              NaN   
25720              NaN  NaN  ...              NaN   
25721              NaN  NaN  ...              NaN   
25722              NaN  NaN  ...              NaN   

               Investigator or Source                       pdf  \
0                    R. Collier, GSAF      2018.06.25-Wolfe.pdf   
1      K.McMurray, TrackingSharks.com    2018.06.18-McNeely.pdf   
2      K.McMurray, TrackingSharks.com     2018.06.09-Denges.pdf   
3                      B. Myatt, GSAF  2018.06.08-Arrawarra.pdf   
4                           A .Kipper      2018.06.04-Ramos.pdf   
...                               ...                       ...   
25718                             NaN                       NaN   
25719                             NaN                       NaN   
25720                             NaN                       NaN   
25721                             NaN                       NaN   
25722                             NaN                       NaN   

                                            href formula  \
0      http://sharkattackfile.net/spreadsheets/pdf_di...   
1      http://sharkattackfile.net/spreadsheets/pdf_di...   
2      http://sharkattackfile.net/spreadsheets/pdf_di...   
3      http://sharkattackfile.net/spreadsheets/pdf_di...   
4      http://sharkattackfile.net/spreadsheets/pdf_di...   
...                                                  ...   
25718                                                NaN   
25719                                                NaN   
25720                                                NaN   
25721                                                NaN   
25722                                                NaN   

                                                    href Case Number.1  \
0      http://sharkattackfile.net/spreadsheets/pdf_di...    2018.06.25   
1      http://sharkattackfile.net/spreadsheets/pdf_di...    2

## Sharks Information

In [221]:
sharks.info()   

"""Con esta función recibimos información sobre la data frame. 
Número de entradas: 25723
Número de columnas: 24
Número de elementos nulos dependiendo de la columna
El Data Type de la información en cada columna: float64(2), object(22)
Memory usage: 4.7+ MB
"""


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8702 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null 

'Con esta función recibimos información sobre la data frame. \nNúmero de entradas: 25723\nNúmero de columnas: 24\nNúmero de elementos nulos dependiendo de la columna\nEl Data Type de la información en cada columna: float64(2), object(22)\nMemory usage: 4.7+ MB\n'

With this function, we receive information about the data frame.

Range Index: 25723 entries, 0 to 25722

Number of entries: 25723

Number of columns: 24

Number of null elements depending on the column

The Data Type of the information in each column: float64(2), object(22)

Memory usage: 4.7+ MB

#### Number of rows

In [222]:
n_rows = sharks.shape[0]

print(f"The number of rows is {n_rows}")

The number of rows is 25723


#### Number of columns

In [223]:
n_columns = sharks.shape[1]

print(f"The number of columns is {n_columns}")

The number of columns is 24


### Columns

#### Modifying the columns names

In [224]:
sharks.columns = [i.lower().replace(" ", "_").strip() for i in sharks.columns]

# Modify the column names to give them a more readable and unified appearance. 

#This will be helpfull in the future, to search or change columns, or manipulate the information regarding columns

sharks.columns # Now we can see the modified name of the columns

Index(['case_number', 'date', 'year', 'type', 'country', 'area', 'location',
       'activity', 'name', 'sex_', 'age', 'injury', 'fatal_(y/n)', 'time',
       'species_', 'investigator_or_source', 'pdf', 'href_formula', 'href',
       'case_number.1', 'case_number.2', 'original_order', 'unnamed:_22',
       'unnamed:_23'],
      dtype='object')

#### The amount of Null Values per column

In [225]:
pd.isna(sharks).sum()  # The number is the amount of Null values in sharks

case_number               17021
date                      19421
year                      19423
type                      19425
country                   19471
area                      19876
location                  19961
activity                  19965
name                      19631
sex_                      19986
age                       22252
injury                    19449
fatal_(y/n)               19960
time                      22775
species_                  22259
investigator_or_source    19438
pdf                       19421
href_formula              19422
href                      19421
case_number.1             19421
case_number.2             19421
original_order            19414
unnamed:_22               25722
unnamed:_23               25721
dtype: int64

#### Not missing values per relevant column

In [226]:
sharks.rename(columns={'fatal_(y/n)': 'fatal'}, inplace= True) # changing the name of Fatal

sharks[['fatal']] # checking if the change was done

# I've been having problems with the name of this column so I'm using .rename() to make it simpler

fatal
0         N
1         N
2         N
3         N
4         N
...     ...
25718   NaN
25719   NaN
25720   NaN
25721   NaN
25722   NaN

[25723 rows x 1 columns]

### Checking the not missing values by Column

In [227]:
list_relevant_columns = ["year", "country", "sex_", "type", "activity", "injury", "fatal"] 

# I'm creating a variable with the list of the relevant columns to access easier to it in next cells

list_relevant_columns

['year', 'country', 'sex_', 'type', 'activity', 'injury', 'fatal']

#### Using a For Loop to access the not missing values in each relevant Column

In [228]:
# To reach the hypothesis I want to prove I want to have an idea of the relevant information that I have access in Sharks
# So I'm going to check the total amount of not missing values in the columns

notmissing_counts = sharks[list_relevant_columns].count() 

# I have created a variable "notmissing_counts" that will have all the counts from the relevant columns of Sharks

for column, count in notmissing_counts.items():
    print(f"The number of not missing values per {column} is {count}")
    
# With this for loop I'm iteratin all the items inside the variable "notmissing_counts" to reach the number of not missing values

The number of not missing values per year is 6300
The number of not missing values per country is 6252
The number of not missing values per sex_ is 5737
The number of not missing values per type is 6298
The number of not missing values per activity is 5758
The number of not missing values per injury is 6274
The number of not missing values per fatal is 5763


I can see an important difference between the relevant information (6300-5737) and the total entries (25723), that I will 
have to erase in the "Cleaning" part of the project

### Different content per relevant Column

In [229]:
list_relevant_columns = ["year", "country", "sex_", "type", "activity", "injury", "age", "fatal"] 

list_relevant_columns

# Now I want to have an idea of the different possibilities of each relevant column.
# For example how many different countries are in Sharks, and if they are correctly introduce

['year', 'country', 'sex_', 'type', 'activity', 'injury', 'fatal']

### Filter to know how many different entries per column and its length 

#### Different Unique values per Year

In [230]:
sharks["year"].unique()

# In this case we can see the unique results inside the column "year", using the function ".unique()"
# This will help me to know the extension of each posible information per column
# The year column, covers from 2018 to maybe 1543, this inconsistence in the years, maybe will be a reason to 
# shorten the number of relevant years for the study
# Googling about, this the first encounters with hawaian surfers were since 1778, so maybe it will be a first start to delete
# I will have to transform the results into ints, not floats

array([2018., 2017.,   nan, 2016., 2015., 2014., 2013., 2012., 2011.,
       2010., 2009., 2008., 2007., 2006., 2005., 2004., 2003., 2002.,
       2001., 2000., 1999., 1998., 1997., 1996., 1995., 1984., 1994.,
       1993., 1992., 1991., 1990., 1989., 1969., 1988., 1987., 1986.,
       1985., 1983., 1982., 1981., 1980., 1979., 1978., 1977., 1976.,
       1975., 1974., 1973., 1972., 1971., 1970., 1968., 1967., 1966.,
       1965., 1964., 1963., 1962., 1961., 1960., 1959., 1958., 1957.,
       1956., 1955., 1954., 1953., 1952., 1951., 1950., 1949., 1948.,
       1848., 1947., 1946., 1945., 1944., 1943., 1942., 1941., 1940.,
       1939., 1938., 1937., 1936., 1935., 1934., 1933., 1932., 1931.,
       1930., 1929., 1928., 1927., 1926., 1925., 1924., 1923., 1922.,
       1921., 1920., 1919., 1918., 1917., 1916., 1915., 1914., 1913.,
       1912., 1911., 1910., 1909., 1908., 1907., 1906., 1905., 1904.,
       1903., 1902., 1901., 1900., 1899., 1898., 1897., 1896., 1895.,
       1894., 1893.,

#### Different Unique values per Country

In [231]:
sharks["country"].unique()

# With this result I can see that there is a mix of Countries, seas, 
# even cities, so I will have to clean it in the Cleaning step

array(['USA', 'AUSTRALIA', 'MEXICO', 'BRAZIL', 'ENGLAND', 'SOUTH AFRICA',
       'THAILAND', 'COSTA RICA', 'MALDIVES', 'BAHAMAS', 'NEW CALEDONIA',
       'ECUADOR', 'MALAYSIA', 'LIBYA', nan, 'CUBA', 'MAURITIUS',
       'NEW ZEALAND', 'SPAIN', 'SAMOA', 'SOLOMON ISLANDS', 'JAPAN',
       'EGYPT', 'ST HELENA, British overseas territory', 'COMOROS',
       'REUNION', 'FRENCH POLYNESIA', 'UNITED KINGDOM',
       'UNITED ARAB EMIRATES', 'PHILIPPINES', 'INDONESIA', 'CHINA',
       'COLUMBIA', 'CAPE VERDE', 'Fiji', 'DOMINICAN REPUBLIC',
       'CAYMAN ISLANDS', 'ARUBA', 'MOZAMBIQUE', 'FIJI', 'PUERTO RICO',
       'ITALY', 'ATLANTIC OCEAN', 'GREECE', 'ST. MARTIN', 'FRANCE',
       'PAPUA NEW GUINEA', 'TRINIDAD & TOBAGO', 'KIRIBATI', 'ISRAEL',
       'DIEGO GARCIA', 'TAIWAN', 'JAMAICA', 'PALESTINIAN TERRITORIES',
       'GUAM', 'SEYCHELLES', 'BELIZE', 'NIGERIA', 'TONGA', 'SCOTLAND',
       'CANADA', 'CROATIA', 'SAUDI ARABIA', 'CHILE', 'ANTIGUA', 'KENYA',
       'RUSSIA', 'TURKS & CAICOS', 'UNITE

#### Length of results by Country

In [232]:
len(sharks["country"].unique())



213

##### Sorting the results by name

In [233]:
# Now I want to see if there are orthographic mistakes or duplicate countries, so I'm going to try sort the results by name

unique_countries = sharks["country"].unique().tolist() # First I'm going to create a variable to transform it into a list

unique_countries_sorted = sorted(unique_countries) # Then using the sorted() function, order them by name

unique_countries_sorted


TypeError: '<' not supported between instances of 'float' and 'str'

In [234]:
# After search about this error I have found that I need to take out the NaN results, so in this way I can, finally, 
# transform and sort the list, to see the problems more accurately

unique_countries = sharks["country"].dropna().unique().tolist()

unique_countries_sorted = sorted(unique_countries)

unique_countries_sorted

# I can see that many of the duplicates come from a space after and before the name of the country, so will have to clean them
# I'm deciding that I'll use the 5-10 countries with more shark attacks
# I'll have to erase continents?

[' PHILIPPINES',
 ' TONGA',
 'ADMIRALTY ISLANDS',
 'AFRICA',
 'ALGERIA',
 'AMERICAN SAMOA',
 'ANDAMAN / NICOBAR ISLANDAS',
 'ANDAMAN ISLANDS',
 'ANGOLA',
 'ANTIGUA',
 'ARGENTINA',
 'ARUBA',
 'ASIA?',
 'ATLANTIC OCEAN',
 'AUSTRALIA',
 'AZORES',
 'BAHAMAS',
 'BAHREIN',
 'BANGLADESH',
 'BARBADOS',
 'BAY OF BENGAL',
 'BELIZE',
 'BERMUDA',
 'BRAZIL',
 'BRITISH ISLES',
 'BRITISH NEW GUINEA',
 'BRITISH VIRGIN ISLANDS',
 'BRITISH WEST INDIES',
 'BURMA',
 'Between PORTUGAL & INDIA',
 'CANADA',
 'CAPE VERDE',
 'CARIBBEAN SEA',
 'CAYMAN ISLANDS',
 'CENTRAL PACIFIC',
 'CEYLON',
 'CEYLON (SRI LANKA)',
 'CHILE',
 'CHINA',
 'COLUMBIA',
 'COMOROS',
 'COOK ISLANDS',
 'COSTA RICA',
 'CRETE',
 'CROATIA',
 'CUBA',
 'CURACAO',
 'CYPRUS',
 'Coast of AFRICA',
 'DIEGO GARCIA',
 'DJIBOUTI',
 'DOMINICAN REPUBLIC',
 'ECUADOR',
 'EGYPT',
 'EGYPT ',
 'EGYPT / ISRAEL',
 'EL SALVADOR',
 'ENGLAND',
 'EQUATORIAL GUINEA / CAMEROON',
 'FALKLAND ISLANDS',
 'FEDERATED STATES OF MICRONESIA',
 'FIJI',
 'FRANCE',
 'FRENCH PO

##### Finding the top 10 Countries by Shark Attacks

In [235]:
top_countries = sharks["country"].value_counts().head(10)

print(top_countries)

# For this I'm using la función ".value_counts()" over the column "Country" to see how many entries has, and then taking the 10 
# first
# Being this 10 the countries with more attacks and with a really big difference bettween them, this result support my Hypothesis
# that the USA is the country with most attack, being almost double
# I'll focus the changes in this 10 countries.

country
USA                 2229
AUSTRALIA           1338
SOUTH AFRICA         579
PAPUA NEW GUINEA     134
NEW ZEALAND          128
BRAZIL               112
BAHAMAS              109
MEXICO                89
ITALY                 71
FIJI                  62
Name: count, dtype: int64


#### Different Unique values per Sex

In [236]:
sharks["sex_"].unique()

# After this search I can see that the entries are not only "Female" and  "Male", so I will have to clean this in the next step
# I'll take as valed F and M
# The rest, I'll unify all of them in a third category named 'unknown' 

array(['F', 'M', nan, 'M ', 'lli', 'N', '.'], dtype=object)

##### Finding the most attacked sex

In [237]:
count_sex = sharks["sex_"].value_counts()

count_sex

# Again this count of entries support the idea that men are more attacked than women with a huuuge difference

sex_
M      5094
F       637
M         2
N         2
lli       1
.         1
Name: count, dtype: int64

#### Different Unique values per Type of attack

In [240]:
sharks["type"].unique()

# In this case same I'll unify all of them in three categories
# "Provoked", "Unprovoked" and "Invalid"


array(['Boating', 'Unprovoked', 'Invalid', 'Provoked', 'Questionable',
       'Sea Disaster', nan, 'Boat', 'Boatomg'], dtype=object)

##### Finding the most common type of attack.

In [241]:
count_type = sharks["type"].value_counts()

count_type

type
Unprovoked      4595
Provoked         574
Invalid          547
Sea Disaster     239
Boating          203
Boat             137
Questionable       2
Boatomg            1
Name: count, dtype: int64

#### Different Unique values per Activity

In [242]:
sharks["activity"].unique()



array(['Paddling', 'Standing', 'Surfing', ...,
       'Crew swimming alongside their anchored ship',
       '4 men were bathing', 'Wreck of  large double sailing canoe'],
      dtype=object)

In [243]:
count_activity = sharks["activity"].value_counts().head(30)

count_activity

# Here I've chosen 30 results because I can see that there are duplicates with names similar but not exactly the same
# So in the Cleaning step I'll have to modify them to add results under an unified option.
# For example: Unify all the surfing results as "surf skiing", "surf fishing", under the same name "surfing"
# And reduce all the entries in the 10 most common activities


activity
Surfing            971
Swimming           869
Fishing            431
Spearfishing       333
Bathing            162
Wading             149
Diving             127
Standing            99
Snorkeling          89
Scuba diving        76
Body boarding       61
Body surfing        49
Swimming            47
Kayaking            33
Treading water      32
Pearl diving        32
Fell overboard      32
Free diving         29
Boogie boarding     29
Windsurfing         19
Walking             17
Boogie Boarding     16
Shark fishing       15
Floating            14
Fishing             13
Canoeing            13
Surf skiing         12
Surf fishing        12
Surf-skiing         12
Rowing              12
Name: count, dtype: int64

#### Different Unique values per Injury

In [244]:
sharks["injury"].unique()

array(['No injury to occupant, outrigger canoe and paddle damaged',
       'Minor injury to left thigh',
       'Injury to left lower leg from surfboard skeg', ...,
       'FATAL, leg stripped of flesh  ',
       'FATAL, knocked overboard by tail of shark & carried off by shark ',
       'FATAL. "Shark bit him in half, carrying away the lower extremities" '],
      dtype=object)

In [245]:
len(sharks["injury"].unique())

3738

In [246]:
count_injury = sharks["injury"].value_counts().head(30)

count_injury

# The results in this case are too messy and caotic, in around 6000 entries there are around 3000, all different including
# leg bitten, what foot was bitten. I think is too chaotic and too specific for the kind of information I want to demonstrate
# so finally I'll take out this column aswell.

injury
FATAL                          802
Survived                        97
Foot bitten                     87
No injury                       82
Leg bitten                      72
Left foot bitten                50
No details                      43
Right foot bitten               39
No injury, board bitten         31
Hand bitten                     29
Thigh bitten                    27
FATAL, body not recovered       24
Calf bitten                     22
Minor injury                    22
Foot lacerated                  21
Right leg bitten                20
Lacerations to foot             20
Arm bitten                      20
Lacerations to right foot       18
Lacerations to left foot        18
Ankle bitten                    17
Right calf bitten               16
Heel bitten                     14
No injury to occupants          14
Minor injuries                  14
No injury, surfboard bitten     13
Left leg bitten                 13
Foot severed                    13
Left arm bitt

#### Different Unique values per Age

In [248]:
sharks["age"].unique()

# Here I can see that the results are not clear sometimes, being ranges not defined, I'm interested in the numerical value
# so I'll try to find the mean and apply it 

array(['57', '11', '48', nan, '18', '52', '15', '12', '32', '10', '21',
       '34', '30', '60', '33', '29', '54', '41', '37', '56', '19', '25',
       '69', '38', '55', '35', '46', '45', '14', '40s', '28', '20', '24',
       '26', '49', '22', '7', '31', '17', '40', '13', '42', '3', '8',
       '50', '16', '82', '73', '20s', '68', '51', '39', '58', 'Teen',
       '47', '61', '65', '36', '66', '43', '60s', '9', '72', '59', '6',
       '27', '64', '23', '71', '44', '62', '63', '70', '18 months', '53',
       '30s', '50s', 'teen', '77', '74', '28 & 26', '5', '86', '18 or 20',
       '12 or 13', '46 & 34', '28, 23 & 30', 'Teens', '36 & 26',
       '8 or 10', '84', '\xa0 ', ' ', '30 or 36', '6½', '21 & ?', '75',
       '33 or 37', 'mid-30s', '23 & 20', ' 30', '7      &    31', ' 28',
       '20?', "60's", '32 & 30', '16 to 18', '87', '67', 'Elderly',
       'mid-20s', 'Ca. 33', '74 ', '45 ', '21 or 26', '20 ', '>50',
       '18 to 22', 'adult', '9 & 12', '? & 19', '9 months', '25 to 35',
  

#### Different Unique values per Fatal

In [174]:
sharks["fatal"].unique()

# I can see that there is more than two introduce options, so I'll modify them to unify it in three categories
# "Y", "N" and Unknown

array(['N', 'Y', nan, 'M', 'UNKNOWN', '2017', ' N', 'N ', 'y'],
      dtype=object)

In [247]:
count_fatal = sharks["fatal"].value_counts()

count_fatal

# Anyways, before the modification, we can see that there is a prevalence of non-fatal attacks

fatal
N          4293
Y          1388
UNKNOWN      71
 N            7
M             1
2017          1
N             1
y             1
Name: count, dtype: int64